content/post/deduplicating_pga_with_apollo.md

```diff
@@ -0,0 +1,387 @@
+| Files in PGA          | 1,673,172 | 4,021,258 | 5,549,381  | 909,935 | 1,962,502 |
+| % of all PGA          | 2.47      | 5.93      | 8.18       | 1.34    | 2.89      |
+| Processed             | 1,021,687 | 1,848,084 | 3,066,721  | 628,081 | 1,305,344 |
+| Ratio processed/PGA, %| 61.06     | 45.95     | 55.26      | 69.02   | 66.51     |
+
+{{% caption src="/post/deduplicating_pga_with_apollo/languages.png" %}}
+Percent of files in each language in our corpus.
+{{% /caption %}}
+
+We observed that over 65% of distinct features were **literals** due high variance,
+and apparently were relevant enough for most of the files to be retained. Similarly, 
+due to the quantization, the number of distinct **children** features was 
+minuscule compared to the rest.
+
+|                 | identifiers | literals | graphlets | children | uast2seq | node2vec |
+|-----------------|-------------|----------|-----------|----------|----------|----------|
+| Ratio to all, % | 9.93        | 65.70    | 11.84     | 0.02     | 10.49    | 2.02     |
+
+The average number of features per file seemed
+roughly stable for all the considered languages. Features with the most semantic 
+information, `uast2seq` and `node2vec`, had the highest count on average.
+Note: their numbers depend on the chosen hyperparameters, see the previous section.
+
+|              | identifiers | literals | graphlets | children | uast2seq | node2vec |
+|--------------|-------------|----------|-----------|----------|----------|----------|
+| Python       | 60.71       | 36.20    | 134.55    | 51.90    | 432.73   | 591.61   |
+| Java         | 56.94       |  5.17    |  94.11    | 39.02    | 266.93   | 489.04   |
+| Javascript   | 58.21       | 57.11    | 139.24    | 32.77    | 387.57   | 449.16   |
+| Ruby         | 37.78       | 13.97    |  48.80    | 23.36    | 151.13   | 193.24   |
+| Go           | 80.89       | 49.41    | 110.96    | 41.23    | 326.09   | 467.81   |
+| All languages| 60.41       | 37.59    | 116.23    | 37.37    | 336.30   | 459.77   |
+
+## Maneuvering
+
+The rest of the apollo pipeline ran much smoother. Thanks to [MinhashCuda](https://github.com/src-d/minhashcuda)
+the hashing stage took less then a day to finish. Our previous [blog post](../minhashcuda)
+can provide more details about MinhashCuda. It particularly describes the
+Locality Sensitive Hashing procedure in depth - the same that we employ for the deduplication.
+In a few words, we scan the buckets of the hash tables and treat the files
+in the same bucket as similar and those files form a clique.
+Hence the computational complexity is linear.
+See also [ekzhu/datasketch](https://ekzhu.github.io/datasketch/lsh.html).
+
+We decided to use two similarity thresholds for the hashing process, a stricter 95% 
+and looser 80%, the same the *DéjàVu* authors had used in `SourcersCC`. They yielded
+the following results after the pairwise similarity graph was split into connected
+components (CCs):
```

Comment:
Since the rest of the post focuses on analyzing the CCs and the above paragraph only mentioned briefly how to obtain cliques, it would be good to add a few words about obtaining CCs and re-linking to the appropriate blog entry.

Reply:
updated the paragraph above bc it made no sense to talk only about cliques, as they are not our main focus